media-proxy/src/session-mtl-st2110-20.cc

```diff
@@ -0,0 +1,416 @@
+    return s->query_ext_frame_cb(ext_frame, meta);
+}
+
+static int tx_frame_done_callback_wrapper(void *priv, struct st_frame *frame)
+{
+    tx_st20_mtl_session *s = (tx_st20_mtl_session *)priv;
+    return s->frame_done_cb(frame);
+}
+#endif
+
+void rx_st20_mtl_session::frame_thread()
+{
+    INFO("%s, start\n", __func__);
+    while (!stop) {
+        struct st_frame *frame = st20p_rx_get_frame(handle);
+        if (!frame) { /* no frame */
+            pthread_mutex_lock(&wake_mutex);
+            if (!stop)
+                pthread_cond_wait(&wake_cond, &wake_mutex);
+            pthread_mutex_unlock(&wake_mutex);
+            continue;
+        }
+        /* Verify frame data status. */
+        if (frame->status == ST_FRAME_STATUS_CORRUPTED) {
+            ERROR("[DBG] Received corrupted frame.\n");
+        } else {
+            consume_frame(frame);
+        }
+        st20p_rx_put_frame(handle, frame);
+    }
+}
+
+/* Create new RX session */
+rx_st20_mtl_session::rx_st20_mtl_session(mtl_handle dev_handle, struct st20p_rx_ops *opts,
+                                         memif_ops_t *memif_ops)
+    : handle(0), frame_thread_handle(0), shm_bufs(0), fb_recv(0)
+{
+    if (!dev_handle || !opts || !memif_ops) {
+        cleanup();
+        throw invalid_argument("Input parameters cannot be null");
+    }
+
+    st = dev_handle;
+
+    struct st20p_rx_ops ops_rx = {0};
+    mtl_memcpy(&ops_rx, opts, sizeof(struct st20p_rx_ops));
+
+    ops_rx.priv = this; // app handle register to lib
+    ops_rx.notify_frame_available = frame_available_callback_wrapper;
+
+#if defined(MTL_ZERO_COPY)
+    output_fmt = ops_rx.output_fmt;
+    if (st_frame_fmt_equal_transport(ops_rx.output_fmt, ops_rx.transport_fmt)) {
+        ops_rx.flags |= ST20P_RX_FLAG_EXT_FRAME;
+        ops_rx.query_ext_frame = query_ext_frame_callback_wrapper;
+        ops_rx.flags |= ST20P_RX_FLAG_RECEIVE_INCOMPLETE_FRAME;
+    } else {
+        cleanup();
+        throw invalid_argument("output_fmt and transport_fmt differ");
+    }
+#endif
+
+    handle = st20p_rx_create(dev_handle, &ops_rx);
+    if (!handle) {
+        cleanup();
+        throw runtime_error("Failed to create MTL RX ST20 session");
+    }
+
+    frame_size = st20p_rx_frame_size(handle);
+
+    /* initialize share memory */
+    shm_bufs = (memif_buffer_t *)malloc(sizeof(memif_buffer_t));
+    if (!shm_bufs) {
+        cleanup();
+        throw runtime_error("Failed to allocate memory");
+    }
+
+    int ret = shm_init(memif_ops, frame_size, 2);
+    if (ret < 0) {
+        cleanup();
+        throw runtime_error("Failed to initialize shared memory");
+    }
+
+    /* Start MTL session thread. */
+    frame_thread_handle = new std::thread(&rx_st20_mtl_session::frame_thread, this);
+    if (!frame_thread_handle) {
+        cleanup();
+        throw runtime_error("Failed to create thread");
+    }
+}
+
+void rx_st20_mtl_session::cleanup()
+{
+    if (frame_thread_handle) {
+        frame_thread_handle->join();
```

Comment:
I think it won't cancel the thread. It would wait indefinitely. The stop variable has to be set for this thread to stop.

Reply:
The same issue exists in different places

Reply:
stop variable is set in session_stop()

Reply:
but session_stop() is called from proxy_context.cc, and this join() will freeze the thread that called rx_st20_mtl_session(), which is the thread from proxy_context.cc which would call session_stop() otherwise.

Anyway, you are right. We need to leverage RAII as much as possible, we should move to destructors every cleanup that can be moved there.

Reply:
In this case, to cleanup when constructor fails and throws an exception we could:
1. Use smart pointers whenever you allocate memory
2. Do something like this when smart pointers aren't enough:
```
class Resource {
    // Your resource management logic
public:
    Resource() {
        // Acquire the resource
    }
    ~Resource() {
        // Release the resource
    }
};

class Session{
    Resource resource;  // RAII type
public:
    Session() {
        // If an exception is thrown here, `resource` will still be cleaned up
        if (/* some condition */) {
            throw std::runtime_error("Initialization failed");
        }
    }

    ~Session() {
        // `Resource` will automatically clean itself up
    }
};
```

Reply:
cleanup() no longer exist, destructor se stop variable and join threads